src/Validation/ScannableURLProvider.php

```diff
@@ -0,0 +1,305 @@
+<?php
+/**
+ * Provides URLs to scan.
+ *
+ * @package AMP
+ * @since 2.1
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use AMP_Theme_Support;
+use WP_Query;
+
+/**
+ * ScannableURLProvider class.
+ *
+ * @since 2.1
+ */
+final class ScannableURLProvider {
+
+	/**
+	 * Whether to include URLs that don't support AMP.
+	 *
+	 * @var bool
+	 */
+	private $include_unsupported;
+
+	/**
+	 * An allowlist of conditionals to use for querying URLs.
+	 *
+	 * Usually, this class will query all of the templates that don't have AMP disabled. This allows inclusion based on only these conditionals.
+	 *
+	 * @var string[]
+	 */
+	private $include_conditionals;
+
+	/**
+	 * The maximum number of URLs to provide for each content type.
+	 *
+	 * Templates are each a separate type, like those for is_category() and is_tag(), and each post type is a type.
+	 *
+	 * @var int
+	 */
+	private $limit_per_type;
+
+	/**
+	 * Class construÎctor.
+	 *
+	 * @param integer $limit_per_type       The maximum number of URLs to validate for each type.
+	 * @param array   $include_conditionals An allowlist of conditionals to use for validation.
+	 * @param boolean $include_unsupported  Whether to include URLs that don't support AMP.
+	 */
+	public function __construct(
+		$limit_per_type = 20,
+		$include_conditionals = [],
+		$include_unsupported = false
+	) {
+		$this->limit_per_type       = $limit_per_type;
+		$this->include_conditionals = $include_conditionals;
+		$this->include_unsupported  = $include_unsupported;
+	}
+
+	/**
+	 * Provides the array of URLs to check.
+	 *
+	 * Each URL is an array with two elements, with the URL at index 0 and the type at index 1.
+	 *
+	 * @param int|null $offset Optional. The number of URLs to offset by, where applicable. Defaults to 0.
+	 * @return array Array of URLs and types.
+	 */
+	public function get_urls( $offset = 0 ) {
+		$urls = [];
+
+		/*
+		 * If 'Your homepage displays' is set to 'Your latest posts', include the homepage.
+		 */
+		if ( 'posts' === get_option( 'show_on_front' ) && $this->is_template_supported( 'is_home' ) ) {
+			$urls[] = [
+				'url'  => home_url( '/' ),
+				'type' => 'home',
+			];
+		}
+
+		$amp_enabled_taxonomies = array_filter(
+			get_taxonomies( [ 'public' => true ] ),
+			[ $this, 'does_taxonomy_support_amp' ]
+		);
+		$public_post_types      = get_post_types( [ 'public' => true ], 'names' );
```

Comment:
The `$output` parameter for `get_post_types()` is `'names'` by default.

```suggestion
		$public_post_types      = get_post_types( [ 'public' => true ] );
```